From dd47a5217abcaf6d9085a44e9c1dd8821baf565f Mon Sep 17 00:00:00 2001
From: Gerald Combs <gerald@wireshark.org>
Date: Mon, 28 Mar 2016 15:46:33 -0700
Subject: [PATCH 113/117] SPOOLSS: Try to avoid an infinite loop.

Use tvb_reported_length_remaining in dissect_spoolss_uint16uni. Make
sure our offset always increments in dissect_spoolss_keybuffer.

Conflicts:
	epan/dissectors/packet-dcerpc-spoolss.c

Change-Id: I7017c9685bb2fa27161d80a03b8fca4ef630e793
Reviewed-on: https://code.wireshark.org/review/14687
Reviewed-by: Gerald Combs <gerald@wireshark.org>
Petri-Dish: Gerald Combs <gerald@wireshark.org>
Tested-by: Petri Dish Buildbot <buildbot-no-reply@wireshark.org>
Reviewed-by: Michael Mann <mmann78@netscape.net>
(cherry picked from commit b4d16b4495b732888e12baf5b8a7e9bf2665e22b)
Reviewed-on: https://code.wireshark.org/review/15248
(cherry picked from commit 9753ea8db4ea2caeb3365e842e904cbe5681b79e)
Reviewed-on: https://code.wireshark.org/review/15249
(cherry picked from commit 80006b0eb062a45479d109796df14c13ca7c5785)
Reviewed-on: https://code.wireshark.org/review/16119
Reviewed-by: Balint Reczey <balint@balintreczey.hu>
---
 epan/dissectors/packet-dcerpc-spoolss.c | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/epan/dissectors/packet-dcerpc-spoolss.c b/epan/dissectors/packet-dcerpc-spoolss.c
index 8cd5419..f6bbea8 100644
--- a/epan/dissectors/packet-dcerpc-spoolss.c
+++ b/epan/dissectors/packet-dcerpc-spoolss.c
@@ -1077,7 +1077,7 @@ dissect_spoolss_uint16uni(tvbuff_t *tvb, int offset, packet_info *pinfo _U_,
 
 	/* Get remaining data in buffer as a string */
 
-	remaining = tvb_length_remaining(tvb, offset);
+	remaining = tvb_reported_length_remaining(tvb, offset);
 	if (remaining <= 0) {
 		if (data)
 			*data = g_strdup("");
@@ -6556,9 +6556,10 @@ dissect_spoolss_keybuffer(tvbuff_t *tvb, int offset, packet_info *pinfo,
 		end_offset = tvb_reported_length_remaining(tvb, offset) + 1;
 	}
 
-	while (offset < end_offset)
+	while (offset > 0 && offset < end_offset) {
 		offset = dissect_spoolss_uint16uni(
 			tvb, offset, pinfo, tree, drep, NULL, "Key");
+	}
 
 	return offset;
 }
-- 
2.1.4

